### PR TITLE
Rewrite in ES5

### PR DESCRIPTION
**Pull request by svozza**
As discussed in #3. I wasn't sure what to do with the version number so I bumped it to 2.0.0 even though this isn't a breaking change. Just let me know what you want to do and I'll update the PR.

Also, the slightly verbose way for capturing the arguments comes from here:

https://github.com/petkaantonov/bluebird/wiki/Optimization-killers#3-managing-arguments

Again, if you want me to change it to something like `slice` just say the word.

### CODE REVIEW

**Comment by notxt** [comment]
🏇 🏇 🏇  nice turnaround! I'll read through later today and post any comments I got. It'd be cool to get 🆗's from @nackjicholson and @jinsyaoommen too if they've got bandwidth

File: index.js [hunk 1]
```diff
@@ -1,21 +1,30 @@
...
+ * cannot be handled by a command the command returns undefined. If none of the
+ * supplied commands handle the target, an error is thrown.
  *
  * @param commands
  * @returns {Function}
  */
-function dispatch(...commands) {
+
+module.exports = function dispatch() {
+  var commands = new Array(arguments.length);
```
Comment (nackjicholson) [inline index.js]: Any reason not to use `[].prototype.slice`?

I think that would make it so you don't have to do the loop that assigns `commands[i] = arguments[i]`

``` js
 var commands = [].slice.call(arguments);
```

File: index.js [hunk 2]
```diff
@@ -1,21 +1,30 @@
...
   return function fn(target) {
-    let result;
+    var result;
 
-    for (let command of commands) {
-      result = command(target, fn);
+    for (var j = 0; j < commands.length; j += 1) {
+      result = commands[j](target, fn);
 
       if (!isUndefined(result)) {
```
Comment (nackjicholson) [inline index.js]: What do you think about getting rid of the lodash dependency entirely by using `if (result !== undefined)`?

File: index.js (hunk 1 above)
Comment (svozza) [inline index.js]: Did you have a look at that Bluebird doc I posted? Apparently that leaks the `arguments` object, which causes the V8 engine to deoptimise the function. Tbh, without any perf tests it's probably a premature optimisation.

File: index.js (hunk 2 above)
Comment (svozza) [inline index.js]: Yep, definitely. Do you mind `if (result !== void 0)`?

Comment (nackjicholson) [inline index.js]: :thumbsup:

**Comment by nackjicholson** [comment]
Move `lodash` to a devDependency, and then update the `example` file to use your new es5 version.

This looks pretty close to being able to merge. Thanks a lot for the work

File: index.js (hunk 1 above)
Comment (nackjicholson) [inline index.js]: Nope, I have not looked at that doc. Where is it? Sounds like something the javascript language and v8 needs to fix. Using `Array.prototype.slice.call(arguments);` is used so broadly in all javascript everywhere, and to me is the most straight forward way to handle a variable number of arguments.

Comment (svozza) [inline index.js]: I put the link to the doc in the first message of the PR here but I'll change it if to `slice` if you want, it's no biggie.

**Comment by svozza** [comment]
Think I've addressed all the comments.

**Comment by nackjicholson** [comment]
:ok: :shipit:

**Comment by notxt** [comment]
🆗  thanks @svozza !

**Comment by nackjicholson** [comment]
@notxt Are you planning to publish to npm?

EDIT: Nevermind. I still have access to publishing on npm. Just don't have full git access to the upstream repo

**Pull request merged**